### PR TITLE
mupdf: 1.23.0 -> 1.23.5; only build and install shared libraries

### LINKED ISSUE #89214: clang python bindings
**Project description**
I'm trying to package another program that makes use of clang_9 python bindings: https://github.com/llvm/llvm-project/tree/master/clang/bindings/python

I don't have any experience with clang so I might be wrong, but I think nixpkgs currently doesn't include them. I was expecting them to be copied from src/bindings/python to an output in this expression: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/llvm/9/clang/default.nix

I don't think there are any specific maintainers of llvm for me to ping for help either.

### PR DESCRIPTION
**Pull request by lilyinstarlight**
## Description of changes

Second attempt at #237199

## Things done

- Built on platform(s)
  - [x] x86_64-linux (only direct dependents of mupdf and tested that they function)
  - [x] aarch64-linux (all reverse dependencies via nixpkgs-review and confirmed no regressions against master)
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lilyinstarlight** [comment]
So, uh, it turns out pymupdf is broken on staging independent of this PR after it was bumped in the last python-updates cycle (my test builds were against these commits picked on top of master, so I didn't experience the breakage when running nixpkgs-review)

It's not really related to this PR, but I'll probably still push a fix in the next day or so (but it can go in a new PR since it is orthogonal to this one, mostly)

File: pkgs/desktops/arcan/arcan/default.nix
Comment (AndersonTorres) [inline pkgs/desktops/arcan/arcan/default.nix:24]: I need this patch!

**Comment by lilyinstarlight** [comment]
Okay so I've rebased this to fix a merge conflict, I added support for building the C++ and Python bindings to MuPDF and fixed PyMuPDF to use our own MuPDF build rather than building it's own

I _think_ I've untangled as much MuPDF stuff as I can at this point and I added myself as a maintainer for the package

**Review by teto — APPROVED** [review]

File: pkgs/development/python-modules/libclang/default.nix
Comment (camillemndn) [inline pkgs/development/python-modules/libclang/default.nix:35]: sourceRoot="$sourceRoot/bindings/python"

Comment (camillemndn) [reply pkgs/development/python-modules/libclang/default.nix:35]: Sorry, I was still using LLVM 11 apparently, so this is unnecessary.

Comment (lilyinstarlight) [reply pkgs/development/python-modules/libclang/default.nix:35]: Hmm we won't be using llvm 11 anymore after staging-next is merged, but it looks like at least 12 is still lile that too. I'll adjust it later to support both possible `sourceRoot`s just in case someone wants it

Thank you

**Review by AndersonTorres — APPROVED** [review]

**Review by elohmeier — APPROVED** [review]

**Comment by happysalada** [comment]
Alright I've tried everything I could, so just writing my thoughts here in case someone has an idea.
pymupdf is broken on master because somehow when mupdf builds with ocr enabled, pkg-config is not able to find leptonica.
You can tell the compiler (USE_SYSTEM_LEPTONICA=no) to build with the thirdparty leptonica included, but somehow the build of pymupdf will fail with a circular import error. So I'm guessing that something else needs to be done for the build to actually work.
The best would be to have pkg-config detect leptonica, but I don't understand why it's not happening.
the build will conveniently fail quickly but the error message just points
https://github.com/ArtifexSoftware/mupdf/blob/b426d88c7f77ece0cca739af403d191066357b1f/Makethird#L57
and it's not immediately obvious that leptonica wasn't found.
I'm reaching out here as I'm guessing this is where someone more familiar with the build process than I am might be.
If everyone is busy, no worries of course.

**Comment by lilyinstarlight** [comment]
@happysalada That issue has nothing to do with pkg-config+leptonica (which works just fine if you actually test it) and has nothing to do with this PR. See <https://github.com/NixOS/nixpkgs/pull/338680#issuecomment-2351649366>

**Pull request merged**
